src/consensus/merkle.cpp

```diff
@@ -263,3 +263,296 @@ std::vector<uint256> BlockMerkleBranch(const CBlock& block, uint32_t position)
+}
+
+void swap(MerkleProof& lhs, MerkleProof& rhs)
+{
+    using std::swap;
+    swap(lhs.m_path, rhs.m_path);
+    swap(lhs.m_skip, rhs.m_skip);
+}
+
+MerkleTree::MerkleTree(const MerkleTree& left, const MerkleTree& right)
+{
+    /* Handle the special case of both left and right being fully
+     * pruned, which also results in a fully pruned super-tree.. */
+    if (left.m_proof.m_path.empty() && left.m_proof.m_skip.size()==1 && left.m_verify.empty() &&
+        right.m_proof.m_path.empty() && right.m_proof.m_skip.size()==1 && right.m_verify.empty())
+    {
+        m_proof.m_skip.resize(1);
+        MerkleHash_Sha256Midstate(m_proof.m_skip[0], left.m_proof.m_skip[0], right.m_proof.m_skip[0]);
+        return;
+    }
+
+    /* We assume a well-formed, non-empty MerkleTree for both passed
+     * in subtrees, in which if there are no internal nodes than
+     * either m_skip XOR m_verify must have a single hash. Otherwise
+     * the result of what follows will be an invalid MerkleTree. */
+    m_proof.m_path.emplace_back(MerkleLink::DESCEND, MerkleLink::DESCEND);
+
+    if (left.m_proof.m_path.empty())
+        m_proof.m_path.front().SetLeft(left.m_proof.m_skip.empty()? MerkleLink::VERIFY: MerkleLink::SKIP);
+    m_proof.m_path.insert(m_proof.m_path.end(), left.m_proof.m_path.begin(), left.m_proof.m_path.end());
+    m_proof.m_skip.insert(m_proof.m_skip.end(), left.m_proof.m_skip.begin(), left.m_proof.m_skip.end());
+    m_verify.insert(m_verify.end(), left.m_verify.begin(), left.m_verify.end());
+
+    if (right.m_proof.m_path.empty())
+        m_proof.m_path.front().SetRight(right.m_proof.m_skip.empty()? MerkleLink::VERIFY: MerkleLink::SKIP);
+    m_proof.m_path.insert(m_proof.m_path.end(), right.m_proof.m_path.begin(), right.m_proof.m_path.end());
+    m_proof.m_skip.insert(m_proof.m_skip.end(), right.m_proof.m_skip.begin(), right.m_proof.m_skip.end());
+    m_verify.insert(m_verify.end(), right.m_verify.begin(), right.m_verify.end());
+}
+
+void MerkleTree::clear() noexcept
+{
+    m_proof.clear();
+    m_verify.clear();
+}
+
+void swap(MerkleTree& lhs, MerkleTree& rhs)
+{
+    using std::swap;
+    swap(lhs.m_proof, rhs.m_proof);
+    swap(lhs.m_verify, rhs.m_verify);
+}
+
+uint256 MerkleTree::GetHash(bool* invalid) const
+{
+    std::vector<std::pair<bool, uint256> > stack(2);
+    auto verify_pos = m_verify.begin();
+    auto verify_last = m_verify.end();
+    auto skip_pos = m_proof.m_skip.begin();
+    auto skip_last = m_proof.m_skip.end();
+
+    auto visitor = [&stack, &verify_pos, &verify_last, &skip_pos, &skip_last](std::size_t depth, MerkleLink value, bool right) -> bool
+    {
+        const uint256* new_hash = nullptr;
+        switch(value) {
+            case MerkleLink::DESCEND:
+                stack.emplace_back();
+                return false;
+
+            case MerkleLink::VERIFY:
+                if (verify_pos == verify_last) // read past end of verify hashes list
+                    return true;
```

Comment:
I'm confused here. It looks like the encoded path will have a final "VERIFY" which is actually does not verify anything (but instead terminates the loop successfully). Maybe expand on the comment a bit as to why VERIFY results in no verification made here.

Reply:
I think perhaps `verify_last` is misnamed. It is defined to be `m_verify.end()` above, which would be one slot _past_ the back of the vector. So I don't think there is a terminal / unused "VERIFY" here.